multidimensional-pod-autoscaler/KEP.md

```diff
@@ -0,0 +1,519 @@
+- Recommendation and the actuation are completely separated
+
+Cons:
+- Webhooks introduce extra overhead for vertical scaling operations (can be avoided after in-place resizing of pod is enabled without eviction)
+- Vertical and horizontal scaling executions are separated (can be avoided after in-place resizing of pod is enabled without eviction)
+- State changes in pod sizes are not persisted (too much to keep in etcd, could use Prometheus to store pod state changes)
+
+### Action Recommendation Implementation
+
+To generate the vertical scaling action recommendation, we reuse VPA libraries as much as possible to implement scaling algorithm integrated with the newly generated MPA API code.
+To do that, we need to update accordingly the code which reads and updates the VPA objects to be interacting with the MPA objects.
+To generate the horizontal scaling action recommendation, we reuse HPA libraries, integrating with the MPA API code, to reads and updates the MPA objects.
+We integrate vertical and horizontal scaling in a single feedback cycle.
+As an intitial solution, vertical scaling and horizontal scaling is performed alternatively (vertical scaling first).
+In the future, we can consider more complex way of prioritization and conflict resolution.
+The separation of recommendation and actuation allows customized recommender to be used to replace the default recommender.
+For example, users can plug-in their RL-based controller to replace the MPA recommender, receiving measurements from the Metrics Server and modifying the MPA objects directly to give recommendations.
+
+The implementation of the MPA framework (the backend) is based on the existing HPA and VPA codebase so that it only requires minimum code maintenance.
+Reused Codebase References:
+- HPA: https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/podautoscaler
+- VPA: https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler
+
+### MPA API Object
+
+We reuse the CR definitions from the [MultidimPodAutoscaler](https://cloud.google.com/kubernetes-engine/docs/how-to/multidimensional-pod-autoscaling) object developed by Google.
+`MultidimPodAutoscaler` is the configuration for multi-dimensional Pod autoscaling, which automatically manages Pod resources and their count based on historical and real-time resource utilization.
+MultidimPodAutoscaler has two main fields: `spec` and `status`.
+
+#### MPA Spec
+
+```
+# MultidimPodAutoscalerSpec
+apiVersion: autoscaling.gke.io/v1beta1
+kind: MultidimPodAutoscaler
+metadata:
+  name: my-autoscaler
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: my-target
+  updatePolicy:
+    updateMode: Auto
+  metrics:
+  - type: Resource
+    resource:
+    # Define the target CPU utilization request here
+    name: cpu
+    target:
+      type: Utilization
+      averageUtilization: target-cpu-util
+  constraints:
+    minReplicas: min-num-replicas
+    maxReplicas: max-num-replicas
```

Comment:
I'd expect to see the minimum and maximum memory (or other resources) listed here.

Reply:
The min/max memory is specified under the resource policy field, following the convention defined in the verticalpodautoscaler spec. Let me know if that is confusing.